src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryLogic.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.URI;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+
+public class ConfigurableRetryLogic {
+    private final static int INTERVAL_BETWEEN_READS = 30000;
+    private final static String DEFAULT_PROPS_FILE = "mssql-jdbc.properties";
+    private static final java.util.logging.Logger CONFIGURABLE_RETRY_LOGGER = java.util.logging.Logger
+            .getLogger("com.microsoft.sqlserver.jdbc.ConfigurableRetryLogic");
+    private static ConfigurableRetryLogic driverInstance = null;
+    private static long timeLastModified;
+    private static long timeLastRead;
+    private static String lastQuery = ""; // The last query executed (used when rule is process-dependent)
+    private static String rulesFromConnectionString = "";
+    private static HashMap<Integer, ConfigRetryRule> stmtRules = new HashMap<>();
+    private static final Lock CRL_LOCK = new ReentrantLock();
+
+    private ConfigurableRetryLogic() throws SQLServerException {
+        timeLastRead = new Date().getTime();
+        setUpRules();
+    }
+
+    /**
+     * Fetches the static instance of ConfigurableRetryLogic, instantiating it if it hasn't already been.
+     * Each time the instance is fetched, we check if a re-read is needed, and do so if properties should be re-read.
+     *
+     * @return The static instance of ConfigurableRetryLogic
+     * @throws SQLServerException
+     *         an exception
+     */
+    public static ConfigurableRetryLogic getInstance() throws SQLServerException {
+        CRL_LOCK.lock();
+        try {
+            if (driverInstance == null) {
+                driverInstance = new ConfigurableRetryLogic();
+            } else {
+                reread();
+            }
+            return driverInstance;
+        } finally {
+            CRL_LOCK.unlock();
+        }
+    }

```

Comment:
We should do double checked locking here. Something like below:

```suggestion
    public static ConfigurableRetryLogic getInstance() throws SQLServerException {
        if (driverInstance == null) {
            CRL_LOCK.lock();

            try {
                if (driverInstance == null) {
                    driverInstance = new ConfigurableRetryLogic();
                } else {
                    reread();
                }
            } finally {
                CRL_LOCK.unlock();
            }

        } else {
            reread();
        }

        return driverInstance;
    }
```